File: ParrotSpam.cs
using System;

class ParrotSpam
{
    public const int NUM_CHARS = 4000;

    static int Main(string[] args)
    {
        var parrots = Array.FindAll(args, a => a.Length > 0);
        if (parrots.Length == 0) {
            Console.Error.WriteLine("Nothing to repeat");
            return 1;
        }

        for (int c = 0; ; ) {
            foreach (string p in parrots) {
                c += p.Length;
                if (c > NUM_CHARS)
                    goto Done;

                Console.Write(p);
            }
        }
    Done:
        if (!Console.IsOutputRedirected)
            Console.WriteLine();

        return 0;
    }
}

File: ParrotSpam.pas
program ParrotSpam;

uses
    classes,
    sysutils;

function spam(remaining: Integer; parrots: TStringList): AnsiString;
var
    builder: TStringBuilder;
    i: Integer;
label
    out;
begin
    builder := TStringBuilder.create(remaining);

    while true do
    begin
        for i := 0 to parrots.count - 1 do
        begin
            if remaining < length(parrots[i]) then
                goto out;
            builder.append(parrots[i]);
            remaining -= length(parrots[i]);
        end;
    end;

    out:
    spam := builder.toString;
end;

var
    parrots: TStringList;
    i: Integer;
begin
    parrots := TStringList.create;

    for i := 1 to paramCount() do
        if length(paramStr(i)) > 0 then
            parrots.add(paramStr(i));

    if parrots.count = 0 then
    begin
        writeLn('Nothing to repeat');
        halt(1);
    end;

    writeLn(spam(4000, parrots));
end.
